src/app/modules/main/wallet_section/all_tokens/flat_tokens_model.nim

```diff
@@ -129,26 +130,53 @@ QtObject:
 
 
   proc modelsAboutToUpdate*(self: FlatTokensModel) =
+    self.tokenMarketDetails = @[]
     self.beginResetModel()
 
   proc modelsUpdated*(self: FlatTokensModel) =
-    self.tokenMarketDetails =  @[]
     for token in self.delegate.getFlatTokensList():
-      self.tokenMarketDetails.add(newMarketDetailsItem(self.marketValuesDelegate, token.symbol))
+      let symbol = if token.communityId.isEmptyOrWhitespace: token.symbol
+                   else: ""
+      self.tokenMarketDetails.add(newMarketDetailsItem(self.marketValuesDelegate, symbol))
     self.endResetModel()
 
   proc tokensMarketValuesUpdated*(self: FlatTokensModel) =
-    for i in countup(0, self.rowCount()):
-      let index = self.createIndex(i, 0, nil)
+    if self.delegate.getFlatTokensList().len > 0:
+      let index = self.createIndex(0, 0, nil)
+      let lastindex = self.createIndex(self.delegate.getFlatTokensList().len-1, 0, nil)
+      defer: index.delete
+      defer: lastindex.delete
+      self.dataChanged(index, lastindex, @[ModelRole.MarketDetails.int, ModelRole.MarketDetailsLoading.int])
+
+  proc tokensMarketValuesAboutToUpdate*(self: FlatTokensModel) =
+    if self.delegate.getFlatTokensList().len > 0:
+      let index = self.createIndex(0, 0, nil)
+      let lastindex = self.createIndex(self.delegate.getFlatTokensList().len-1, 0, nil)
+      defer: index.delete
+      defer: lastindex.delete
+      self.dataChanged(index, lastindex, @[ModelRole.MarketDetails.int, ModelRole.MarketDetailsLoading.int])
+
+  proc tokensDetailsAboutToUpdate*(self: FlatTokensModel) =
+    if self.delegate.getFlatTokensList().len > 0:
+      let index = self.createIndex(0, 0, nil)
+      let lastindex = self.createIndex(self.delegate.getFlatTokensList().len-1, 0, nil)
       defer: index.delete
-      self.dataChanged(index, index, @[ModelRole.MarketDetails.int, ModelRole.MarketDetailsLoading.int])
+      defer: lastindex.delete
+      self.dataChanged(index, lastindex, @[ModelRole.Description.int, ModelRole.WebsiteUrl.int, ModelRole.DetailsLoading.int])
 
   proc tokensDetailsUpdated*(self: FlatTokensModel) =
```

Comment:
I can't see any difference between this method and the above (`tokensDetailsAboutToUpdate`); what's the deal here?

Reply:
ok so the tokensDetailsAboutToUpdate is emitted when refetch the details from the service via an async call and tokensDetailsUpdated when the details are fetched. So we update the model accordingly. Here we rely on data from service and dont hold our own copies of the data and hence this is needed. This will also help updating UI state to loading/loaded as needed.

Reply:
Yeah but still... the implementation is the same? That's what looked fishy :)

Reply:
aah yes you are right, we are doing the same thing both places, I could add a function to avoid code repetition but it is intentional